Fast Style Transfer

yining1023 opened the pull request:
This PR -
- adds `TransformNet` class, and `TransformNet.setStyle()` and `TransformNet.predict()` methods
- adds a simple example of Fast Style Transfer with one model about wave style

I will create a demo of Fast Style Transfer with a few other styles and host it on my GitHub, because it is a large folder of all the models with different styles.

You can check out a demo [here](https://yining1023.github.io/fast_style_transfer_in_p5ML/).

Simple example of Fast Style Transfer:
<img width="500" alt="screen shot 2018-01-18 at 2 25 41 am" src="https://user-images.githubusercontent.com/8662372/35085553-326d0d44-fbf7-11e7-9871-0bde42c75773.png">

shiffman commented on examples/fast_style_transfer/sketch.js:
```diff
@@ -0,0 +1,49 @@
+  * @param inputImg HTMLImageElement of input img
+  * @return Array3D containing pixels of output img
+  */
+  outputImgData = net.predict(inputImg);
```
It's tricky b/c I think the library should be able to be "p5 independent" and work with plain JS or other JS frameworks, however, for beginners using p5 it would be nice to be able to pass in a `p5.Image` or valid `p5.Element`/`p5.MediaElement` and receive a `p5.Image` back.

yining1023 replied:
I totally agree. I added another function called `array3DToP5Image(imageData)` which takes an Array3D and returns a p5.Image. I left this `array3DToP5Image` in the example code. I was going to include it in the p5ml library, but in order to do that, I need to add p5.js as a dependency to p5.ml. And I'm not sure if p5ml should depend on p5.js. I think this is a question that worth discussion. So I left the function in the example code for now.

shiffman commented on examples/fast_style_transfer/sketch.js:
```diff
@@ -0,0 +1,49 @@
+  for (let i = 0; i < width; i++) {
+    for (let j = 0; j < height; j++) {
+      k = (i + j * height) * 3;
+      let r = Math.round(255 * data[k + 0]);
```
I believe p5 has a `round()` function so `Math` is not necessary. Also, would `floor()` and `256` be more correct here?  Tiny minor point! And perhaps irrelevant if we refactor based on my other comment above.

yining1023 replied:
You are right!

shiffman commented on examples/fast_style_transfer/sketch.js:
```diff
@@ -0,0 +1,49 @@
+
+function setup() {
+  createCanvas(252, 252);
+  net = new p5ml.TransformNet(modelLoaded, 'udnie', 'models/udnie/');
```
Is there an option to use this with `preload()`? Also I think the convention is to have the callback last?  Is it possible to load more than one style model into the same `p5ml.TransformNet` instance? Is that way a name is required? Would it be simpler to just create multiple `p5ml.TransformNet` instances with different model directories?

yining1023 replied:
It seems like [p5's `preload()`](https://github.com/processing/p5.js/blob/1acf1e332c57c74f9b899cb28742a7a8809d7406/src/core/core.js#L574) only waits for its native methods, like `loadImage()`. I think using a `callback` might be more generic, and works outside of p5.js.

It's possible to load more than one style model in one instance, but it's simpler to create multiple `p5ml.TransformNet` instance, and a name is not required in this case. so it will be `net1 = new p5ml.TransformNet('models/udnie/', callback)` and `net2 = new p5ml.TransformNet('models/wave/', callback)`.

yining1023 commented:
Thank you!! I left comments under each of your comments. About the documentation to train your model with other styles, I want to train one myself and then write something. For now, all models are from deeplearn.js 's demo.

I changed the method to be `net = new p5ml.TransformNet('models/udnie/', callback)`. And added anther style in the example.
<img width="500" alt="screen shot 2018-01-19 at 2 54 33 am" src="https://user-images.githubusercontent.com/8662372/35140297-b1726698-fcc4-11e7-9052-7f2fa9f2c522.png">

shiffman commented:
Looks great to me, thank you! @cvalenzuela any comments before I merge?

yining1023 commented:
I also added an image utility to convert array3d to HTML image element: `let img = p5ml.array3DToImage(_array3d_);` so it can be friendly to beginners and "p5 independent". It's in my last commit: [c313d40](https://github.com/ITPNYU/p5-deeplearn-js/pull/38/commits/c313d4021bb90258a840d41239dd67e134dabc0d).

cvalenzuela commented:
Looks good to me!, I will merge and add a documentation and example page in the website
